### PR TITLE
Don't pile up context_meter callbacks

### LINKED ISSUE #7949: Metrics introduce non-trivial overhead
I recently noticed that our per-task overhead increased to an extend that is not critical but still concerning.

To test this, I ran a very simple workload

```python
    from distributed import Client
    with Client() as c:
        def inc(x):
            return x + 1
        futs = c.map(inc, range(10_000))
```

First run, this shows quite some whitespace

image

but I anticipate that most of this whitespace is due to `worker-saturation` being set and the mapped tasks are actually detected as root tasks. Workers are just starving.

Disabling worker saturation again reveals a bit of a different problem with metrics. Running a py-spy profile indicates that this line is taking up about 5% of CPU time on the main thread (and another ~2-5% on the worker thread)

https://github.com/dask/distributed/blob/e0dbb35e8d4ae12081bb36e51d62008051a9d2c5/distributed/metrics.py#L325

This is not great but the actual problem is that this is also blocking the event loop, increasing all kinds of latencies and overheads as well.

Digging a bit into why this is so expensive, I noticed that the number of callbacks registered here
https://github.com/dask/distributed/blob/e0dbb35e8d4ae12081bb36e51d62008051a9d2c5/distributed/metrics.py#L212-L217

is increasing monotonically for the duration of this compute until it drops at the very end. In this rather trivial example, I get about 1200 callbacks that are evaluated on every `ContextMeter.digest_metric`.

A side-effect of this is that the dashboard much more often shows `deserialize-**` (grey) blocks because [this](https://github.com/dask/distributed/blob/e0dbb35e8d4ae12081bb36e51d62008051a9d2c5/distributed/worker.py#L2205-L2212) typically trivial task deserialization now takes nontrivial time since `_deserialize` is instrumented. (In my example deserializing `inc` supposedly takes 30-120ms)

cc @crusaderky

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #7949 

Reduce overhead of fine performance metrics when there is a very long chain of task finished -> task started events on the worker.

```python
from time import time
import dask
from distributed import Client, wait

dask.config.set({"distributed.scheduler.worker-saturation": float("inf")})
c = Client(n_workers=2, threads_per_worker=2)

def inc(x):
    return x + 1

t0 = time()
N = 10_000
futs = c.map(inc, range(N))
wait(futs)
t1 = time()
print((t1 - t0) / N * c.cluster.scheduler.total_nthreads)
```

| release | time per task |
| --- | --- |
| 2023.3.1 | 1.2ms |
| 2023.3.2 | 6.5ms |
| 2023.6.1 | 8.1ms |
| this PR | 1.5ms |

Note that real-life impact is a lot more modest than what the benchmark would suggest, as you would only experience noticeable degradation when

1. scheduler-side queueing is turned off, or
2. there's a very large amount of non-rootish tasks all ready at the same time

neither of which should be common use cases.

### CODE REVIEW

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -3650,8 +3650,11 @@ def _finish_async_instruction(
             logger.exception("async instruction handlers should never raise!")
             raise
 
-        with ledger.record():
-            # Capture metric events in _transition_to_memory()
+        # Capture metric events in _transition_to_memory()
+        # As this may trigger calls to _start_async_instruction for more tasks,
+        # make sure we don't endlessly pile up context_meter callbacks by specifying
+        # the same key as in _start_async_instruction.
+        with ledger.record(key="async-instruction"):
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:3657]: As an aside, this whole incident would have been prevented by 
- https://github.com/dask/distributed/pull/7686

**Comment by fjetter** [comment]
> scheduler-side queueing is turned off, or
> there's a very large amount of non-rootish tasks all ready at the same time
> neither of which should be common use cases.

Well, this does affect real world problems as well regardless of root-ish classification. For instance, if you add a trivial dependency, nothing is classified as root-ish anymore and you are in the same worst-case situation, e.g.

```python
def inc(x, foo):
    return x + 1

t0 = time()
N = 10_000
foo = c.submit(inc, 0, foo=0, pure=False)
futs = c.map(inc, range(N), foo=foo)
```
(Note: this problem is also suffering a different problem, namely the tasks are all assigned to the same worker. I'm working on this problem in parallel but it also shows the executor overhead)


---


I'm a bit disappointed that this was not detected by our benchmarks suite. I guess our examples are too simple. would you mind adding a parametrized version of [`test_large_map`](https://github.com/coiled/benchmarks/blob/acbcd925451b2508a35be3582235d685247621a7/tests/benchmarks/test_futures.py#L19-L23) that disables queuing. This would've picked up this regression early on.

**Review by fjetter — APPROVED** [review]

**Comment by crusaderky** [comment]
> For instance, if you add a trivial dependency, nothing is classified as root-ish anymore and you are in the same worst-case situation, e.g.
> 
> ```python
> def inc(x, foo):
>     return x + 1
> 
> t0 = time()
> N = 10_000
> foo = c.submit(inc, 0, foo=0, pure=False)
> futs = c.map(inc, range(N), foo=foo)
> ```

You are hitting an edge case, where you should have two TaskGroups but you end up with only one because the dependency and the dependents accidentally share the same prefix AND the keys are not tuples. This causes the TaskGroup to have itself in its dependencies, which in turns causes `is_rootish` to erroneously believe that all 10,001 tasks in the group have 10,001 dependencies.

If you were using dask.array, dask.dataframe, or dask.bag OR the dependency was using a different function, you'd have two different TaskGroups and all the tasks would be rootish (because there are less than 5 dependencies).

We could fix this specific use case by changing Client.map to generate tuple keys.

> I guess our examples are too simple. would you mind adding a parametrized version of [test_large_map](https://github.com/coiled/benchmarks/blob/acbcd925451b2508a35be3582235d685247621a7/tests/benchmarks/test_futures.py#L19-L23) that disables queuing. This would've picked up this regression early on.

I don't think it's a good idea. Queuing impacts way, way more than just this specific issue, so if we actually care the whole test suite should run with and without queuing. I believe there was implicit consensus that that would be overkill as users *shouldn't* tamper with that setting and it's only there to let users revert the change in a hurry in case we realize that we did a catastrophic design mistake that hamstrings a specific use case.

> I'm a bit disappointed that this was not detected by our benchmarks suite. 

I disagree.  I think that the benchmarks suite shows that this is a negligible problem if you don't go digging for edge cases (the big assumption here is that `worker-saturation: inf` is an edge case, and again if it's not meant to be than the whole test suite should run twice).

**Comment by fjetter** [comment]
> If you were using dask.array, dask.dataframe, or dask.bag OR the dependency was using a different function, you'd have two different TaskGroups and all the tasks would be rootish (because there are less than 5 dependencies).

No point in arguing about this. This is a problem, it needs to be fixed. It is very possible to pile up tasks on workers, even with queuing. 

> I don't think it's a good idea. Queuing impacts way, way more than just this specific issue, so if we actually care the whole test suite should run with and without queuing. I believe there was implicit consensus that that would be overkill as users shouldn't tamper with that setting and it's only there to let users revert the change in a hurry in case we realize that we did a catastrophic design mistake that hamstrings a specific use case.

I'm not willing to invest the time and money to run the entire test suite with this being parametrized. This specific test was added to measure task latency and it missed this regression. Disabling this config reinstates this measurement.
